polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp

```diff
@@ -1056,8 +1056,90 @@ struct EraseSPIRVBuiltinPattern
   }
 };
 
-struct ConvertPolygeistToLLVMPass
+class DuplicateFuncCleaner {
+public:
+  struct Entry {
+    Entry(StringRef funcName, FunctionType funcType)
+        : funcName(funcName), signature(funcType) {}
+
+    StringRef funcName;
+    FunctionType signature;
+    bool anyHadConflictingSignature = false;
+    bool found = false;
+  };
+
+  explicit DuplicateFuncCleaner(std::initializer_list<Entry> init)
+      : entries(init) {
+    llvm::sort(entries, [](const Entry &lhs, const Entry &rhs) {
+      return lhs.funcName < rhs.funcName;
+    });
+  }
+
+  LogicalResult run(ModuleOp module) {
+    std::vector<LLVM::LLVMFuncOp> toRemove;
+    const auto getEntry = [this](LLVM::LLVMFuncOp func) {
+      auto iter = llvm::lower_bound(entries, func.getName(),
+                                    [](const Entry &entry, StringRef name) {
+                                      return name < entry.funcName;
+                                    });
+      if (iter == entries.end())
+        return iter;
+      return iter->funcName == func.getName() ? iter : entries.end();
+    };
+    for (auto func : module.getOps<LLVM::LLVMFuncOp>()) {
+      auto iter = getEntry(func);
+      if (iter == entries.end())
+        continue;
+      // If the type is different, annotate it in the entry. This will lead to
+      // an error only if more than one occurrence of the function is found.
+      iter->anyHadConflictingSignature |=
+          iter->signature != func.getFunctionType();
+      // If it was already found, the current function should be removed.
+      // Otherwise, set the function as already found.
+      if (iter->found)
+        toRemove.push_back(func);
+      else
+        iter->found = true;
+    }
+    for (LLVM::LLVMFuncOp func : toRemove) {
+      auto iter = getEntry(func);
```

Comment:
Wouldn't it be more efficient to store the entry together with the function to be removed, so we don't have to traverse the entries multiple times through `getEntry`?

Reply:
Done!